scripts/format_code.py

```diff
@@ -0,0 +1,257 @@
+    recurse: when True, will also recursively search for files in any
+      subdirectories found in path.
+  
+  Returns:
+    A list of matching filenames found in the directory.
+  """
+  filenames = []
+  for root, dirs, files in os.walk(path):
+    path = root.split(os.sep)
+    for filename in files:
+      if filename.endswith(FILE_TYPE_EXTENSIONS):
+        full_path = os.path.join(root, filename)
+        if FLAGS.verbose:
+          print("\t\t" + full_path)
+        filenames.append(full_path)  
+      if not recurse:
+        break;
+  return filenames
+
+def directory_search_list_files():
+  """Create a list of files based on the directories defined in the command
+  line arguments -dr and -d.
+  
+  Returns:
+    A list of filenames which match one of the extensions in
+    FILE_TYPE_EXTENSIONS.
+  """
+  filenames = []
+  if FLAGS.dr:
+    if FLAGS.verbose:
+      print()
+      print("Adding files by recursivly traversing directories:")
+    for directory in FLAGS.dr:
+      if FLAGS.verbose:
+        print("\tRecursive searching directory: \""+directory+"\"")
+      filenames += list_files_from_directory(directory, True)
+  if FLAGS.d:
+    for directory in FLAGS.d:
+      print("\Searching files in directory: \""+directory+"\"")
+      filenames += list_files_from_directory(directory, False)
+  if FLAGS.verbose:
+    print("  Found", len(filenames), "files.")
+    print()
+  return filenames
+
+def validate_arguments():
+  """Ensures that a proper command line configuration exists to create a file
+  list either via git-diff operations or via manuallyd defined --file lists.
+  Logs an error if theres an errant configuration.
+  
+  Returns:
+    True if the either directories or git diff is defined, or both.  Returns
+    False otherwise signallingt hat execution should end.
+  """
+  if not FLAGS.git_diff:
+    if not FLAGS.file: 
+      print()
+      print("ERROR:  -git_diff not defined, and there are no -file targets.")
+      print("Nothing to do. Exiting.")
+      print()
+      return False
+  return True
+
+def main(argv):
+  if not validate_arguments():
+    if FLAGS.verbose:
+        print("Returning exit code 2")
+    sys.exit(2)
+  
+  filenames = []
+  if FLAGS.file:
+      filenames = FLAGS.file
+
+  if FLAGS.dr or FLAGS.d:
+    filenames += directory_search_list_files()
+
+  if FLAGS.git_diff:
+    filenames += git_diff_list_files()
+
+  print("Checking the format of", len(filenames), "file(s). Please wait:")
+  exit_code = 0
+  if FLAGS.format_file:
+    count = 0
+    for filename in filenames:
+        if does_file_need_formatting(filename):
+          print("  - Formatting: \""+filename+"\"")
+          format_file(filename)
+          count += 1
+        else:
+           print("  - File already formatted: \""+filename+"\"")
+    print("  - Formatted", count, "files.")
+  else:
+    count = 0
+    for filename in filenames:      
+      if does_file_need_formatting(filename):  
+        exit_code = 1
```

Comment:
nit: non zero exit code usually indicates command failure but the script doesn't really fail here as we did not explicitly ask for formatting files. 
I am guessing this is for github server side to know that files weren't formatted and we shouldn't really continue. In that case, instead of using exit codes, it might be better to raise a `ValueError` with a nice descriptive message.

```
raise ValueError('something went wrong')
```

Reply:
I wanted to use this script in a git hook, and [the pre commit git hook uses exit codes to signal errors](https://git-scm.com/docs/githooks#_pre_commit), so I'm going to stick witht his current implementation until I can test that out properly.   I assume that I can just check the exit code of the script in GHA as well.

Reply:
Makes sense. Thanks for clarifying!